File: basics/basics-typescript/src/utils/example_stubs.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate Examples for the Node.js/TypeScript SDK,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/blob/main/LICENSE
 */

import { applicationError, maybeCrash } from "./failures"

export type UserRole = {
  roleKey: string,
  roleDescription: string
}
  
export type Permission = {
  permissionKey: string,
  setting: string
}
  
export type UpdateRequest = {
  userId: string,
  role: UserRole,
  permissons: Permission[]
}

export async function getCurrentRole(userId: string): Promise<UserRole> {
  // in this example, the previous role was always just 'viewer'
  return { roleKey: "viewer", roleDescription: "User that cannot do much" }
}

/*
  * This function would call the service or API to record the new user role.
  * For the sake of this example, we just fail with a random probability and
  * otherwise return success.
  */
export async function applyUserRole(userId: string, userRole: UserRole): Promise<boolean> {
  maybeCrash(0.3);
  console.log(`>>> Applied role ${userRole.roleKey} for user ${userId}`);
  return true;
}

/*
  * This function would call the service or API to apply a permission.
  * For the sake of this example, we just fail with a random probability
  * and otherwise return success.
  */
export async function applyPermission(userId: string, permission: Permission): Promise<void> {
  maybeCrash(0.2);
  console.log(`>>> Applied permission ${permission.permissionKey}:${permission.setting} for user ${userId}`);
}

/*
 * This function would typically call the service or API to record the new user role.
 * For the sake of this example, we sometimes fail when applying an advanced role
 * and otherwise return success.
 */
export async function tryApplyUserRole(userId: string, userRole: UserRole): Promise<void> {
  maybeCrash(0.3); // sometimes infra goes away

  if (userRole.roleKey !== "viewer") {
    applicationError(0.3, `Role ${userRole.roleKey} is not possible for user ${userId}`);
  }
  console.log(`>>> Applied role ${userRole.roleKey} for user ${userId}`);
}

/*
 * This function would call the service or API to apply a permission.
 * For the sake of this example, we sometimes fail when applying an 'allow' permission
 * and otherwise return success. Also, we sometimes crash the process.
 */
export async function tryApplyPermission(userId: string, permission: Permission): Promise<Permission> {
  const { permissionKey, setting } = permission;
  maybeCrash(0.3); // sometimes infra goes away

  if (setting !== "blocked") {
    applicationError(0.4, `Could not apply permission ${permissionKey}:${setting} for user ${userId} due to a conflict.`);
  }
  console.log(`>>> Applied permission ${permissionKey}:${setting} for user ${userId}`);
  return { permissionKey, setting: "blocked" }
}

File: basics/basics-java/src/main/java/utils/ExampleStubs.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate Examples for the Node.js/TypeScript SDK,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/blob/main/LICENSE
 */
package utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import static utils.FailureStubs.applicationError;
import static utils.FailureStubs.maybeCrash;

public class ExampleStubs {

    private static final Logger logger = LogManager.getLogger(ExampleStubs.class);

    public static boolean applyUserRole(String userId, UserRole role) {
        maybeCrash(0.3);
        logger.info(String.format(">>> Applying role %s to user %s", role, userId));
        return true;
    }

    public static void applyPermission(String userId, Permission permission) {
        maybeCrash(0.2);
        logger.info(String.format(">>> Applying permission %s:%s for user %s",
                permission.getPermissionKey(),
                permission.getSetting(),
                userId));
    }

    public static UserRole getCurrentRole(String userId){
        return new UserRole("viewer", "User cannot do much");
    }

    public static void tryApplyUserRole(String userId, UserRole role) {
        maybeCrash(0.3);

        if(!role.getRoleKey().equals("viewer")){
            applicationError(0.3, "Role " + role.getRoleKey() + " is not possible for user " + userId);
        }
        logger.error(String.format(">>> Applying role %s to user %s", role, userId));
    }

    public static Permission tryApplyPermission(String userId, Permission permission){
        maybeCrash(0.3);

        if(!permission.getSetting().equals("blocked")) {
            applicationError(0.4,
                    "Could not apply permission " + permission.getPermissionKey() +
                            ":" + permission.getSetting() + " for user" + userId + " due to a conflict."
                    );
        }
        logger.info(String.format(">>> Applying permission %s:%s for user %s",
                permission.getPermissionKey(),
                permission.getSetting(),
                userId));

        return new Permission(permission.getPermissionKey(), "blocked");
    }

    public static String updateUserProfile(String profile) {
        return Math.random() < 0.8 ? "NOT_READY" : profile + "-id";
    }

    public static String setUserPermissions(String userId, String permissions) {
        return permissions;
    }

    public static void provisionResources(String userId, String role, String resources){}

}
